dll/win32/kernel32/client/thread.c

```diff
@@ -1092,240 +1092,136 @@ QueueUserWorkItem(IN LPTHREAD_START_ROUTINE Function,
     return TRUE;
 }
 
-/*
- * @implemented
- */
-DWORD
-WINAPI
-TlsAlloc(VOID)
-{
-    ULONG Index;
-    PTEB Teb;
-    PPEB Peb;
+// Modification by Dibyamartanda Samanta 
 
-    /* Get the PEB and TEB, lock the PEB */
-    Teb = NtCurrentTeb();
-    Peb = Teb->ProcessEnvironmentBlock;
+DWORD WINAPI TlsAlloc( void )
+{
+    
+    
+    PEB * const peb = NtCurrentTeb()->ProcessEnvironmentBlock;
     RtlAcquirePebLock();
-
-    /* Try to get regular TEB slot */
-    Index = RtlFindClearBitsAndSet(Peb->TlsBitmap, 1, 0);
-    if (Index != 0xFFFFFFFF)
-    {
-        /* Clear the value. */
-        Teb->TlsSlots[Index] = 0;
-        RtlReleasePebLock();
-        return Index;
-    }
-
-    /* If it fails, try to find expansion TEB slot. */
-    Index = RtlFindClearBitsAndSet(Peb->TlsExpansionBitmap, 1, 0);
-    if (Index != 0xFFFFFFFF)
+	DWORD index  = RtlFindClearBitsAndSet( peb->TlsBitmap, 1, 1 );
+    if (index != ~0U) 
+	{
+	NtCurrentTeb()->TlsSlots[index] = 0; 
+	}
+    else
     {
-        /* Is there no expansion slot yet? */
-        if (!Teb->TlsExpansionSlots)
-        {
-            /* Allocate an array */
-            Teb->TlsExpansionSlots = RtlAllocateHeap(RtlGetProcessHeap(),
-                                                     HEAP_ZERO_MEMORY,
-                                                     TLS_EXPANSION_SLOTS *
-                                                     sizeof(PVOID));
-        }
-
-        /* Did we get an array? */
-        if (!Teb->TlsExpansionSlots)
-        {
-            /* Fail */
-            RtlClearBits(Peb->TlsExpansionBitmap, Index, 1);
-            Index = 0xFFFFFFFF;
-            BaseSetLastNTError(STATUS_NO_MEMORY);
-        }
-        else
+        index = RtlFindClearBitsAndSet( peb->TlsExpansionBitmap, 1, 0 );
+        if (index != ~0U)
         {
-            /* Clear the value. */
-            Teb->TlsExpansionSlots[Index] = 0;
-            Index += TLS_MINIMUM_AVAILABLE;
+            if (!NtCurrentTeb()->TlsExpansionSlots &&
+                !(NtCurrentTeb()->TlsExpansionSlots = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY,
+                                         8 * sizeof(peb->TlsExpansionBitmapBits) * sizeof(void*) )))
+            {
+                RtlClearBits( peb->TlsExpansionBitmap, index, 1 );
+                index = ~0U;
+                SetLastError( ERROR_NOT_ENOUGH_MEMORY );
```

Comment:
Could you compare this code of yours with the old one and tell us why those parts of the old code had to be changed?

Reply:
I re implemented it , i will compare it and come back

Reply:
And why did you "reimplement" it? It's functionally almost identical and it doesn't fix anything, it rather introduces a bug.